IPython/core/guarded_eval.py

```diff
@@ -817,6 +817,86 @@ def dummy_function(*args, **kwargs):
         if hasattr(value, "__awaited_type__"):
             return value.__awaited_type__
         return value
+    if isinstance(node, ast.While):
+        loop_locals = context.transient_locals.copy()
+        loop_context = context.replace(transient_locals=loop_locals)
+
+        result = None
+        for stmt in node.body:
+            result = eval_node(stmt, loop_context)
+
+        policy = get_policy(context)
+        merged_locals = _merge_dicts_by_key(
+            [loop_locals, context.transient_locals.copy()], policy
+        )
+        context.transient_locals.update(merged_locals)
+
+        return result
+    if isinstance(node, ast.For):
+        try:
+            iterable = eval_node(node.iter, context)
+        except:
```

Comment:
We do not want to catch user interrupts or system exceptions (almost never)

```suggestion
        except Exception:
```